File: src/client/controllers/DebugController.ts
import { Controller, OnInit, OnStart } from "@flamework/core";
import { StateController } from "./StateController";
import { Players } from "@rbxts/services";

@Controller()
export class DebugController implements OnInit, OnStart {
	constructor(private stateController: StateController) {}

	private debugUI: ScreenGui | undefined;
	private stateText: TextLabel | undefined;

	onInit() {
		const playerGui = Players.LocalPlayer.WaitForChild("PlayerGui") as PlayerGui;
		this.debugUI = playerGui.WaitForChild("DebugUI") as ScreenGui;
		this.stateText = this.debugUI.FindFirstChild("State") as TextLabel;
	}

	onStart() {
		task.spawn(() => {
			while (true) {
				task.wait();
				if (this.stateText) this.stateText.Text = this.stateController.getCurrentStateName();
			}
		})
	}
}

File: src/client/controllers/DebugController.lua
local DebugController = {};

function DebugController:init(debugMode)
	local playerGui = game:GetService("Players").LocalPlayer:WaitForChild("PlayerGui");

	local DebugUI = playerGui:WaitForChild("DebugUI");

	DebugController.State = DebugUI:WaitForChild("State");
	
	if (debugMode) then
		print("<DebugController> initialized successfully!")
	end
end

function DebugController:ready()
	DebugController.stateController = require(script.Parent.StateController);
	
	task.spawn(function()
		while true do
			task.wait()
			if (DebugController.State and DebugController.stateController) then
				DebugController.State.Text = "State: " .. DebugController.stateController._currentState;
			end
		end
	end)
end

return DebugController;